File: test_contest/solutions/diamond.py
# Solution to diamond from 2017 UCF HSPT

T = int(input())
for t in range(1, T+1):
    # Read in the grid
    n, m = map(int, input().split())
    g = []
    for i in range(0, n):
        g.append(list(input()))
    # marked[i][j] will be true if grid[i][j] is part of a diamond and false otherwise
    marked = []
    # Initialize marked to all false
    for i in range(0, n):
        cur = []
        for j in range(0, m):
            cur.append(False)
        marked.append(cur);
    # Check for diamonds of each possible size
    for size in range(1, n):
        # Try each possible point for the diamond to start where it is possible for it to fit in bounds
        for i in range(0, n - 2*size + 1):
            for j in range(size, m - size + 1):
                good = True;
                # Check that there is a diamond there
                for k in range(0, size):
                    good &= g[i+k][j+k] == '\\'
                    good &= g[i+size+k][j+size-1-k] == '/'
                    good &= g[i+2*size-1-k][j-1-k] == '\\'
                    good &= g[i+size-1-k][j-size+k] == '/'
                # If a diamond was found, mark all characters that were part of a diamond
                if good:
                    for k in range(0, size):
                        marked[i+k][j+k] = True
                        marked[i+size+k][j+size-1-k] = True
                        marked[i+2*size-1-k][j-1-k] = True
                        marked[i+size-1-k][j-size+k] = True
    # Print the updated grid
    print("Slab #" + str(t) + ":")
    for i in range(0, n):
        for j in range(0, m):
            if marked[i][j]:
                print(g[i][j], end='')
            else:
                print('.', end='')
        print('')
    
        

File: test_contest/solutions/diamond.java

/*
   Solution to HSPT 2017 - Diamonds in the Rough
   Solution Idea - Pick every '/' character as a potential
        starting diamond, then walk down the left half of the diamond
        assuming that you can always make a bigger diamond by going down
        and to the left, or finish a diamond by going straight down seeing
        a '\' character. Be sure to check that the appropriate character
        exists for the right half of the diamond at each step, and when you
        get to the bottom of a candidate diamond, mark all the characters you
        checked as a valid character.
*/

import java.util.Scanner;

public class diamond {
   static int R, C;
   static char[][] map;
   static boolean[][] good;

   public static void main(String[] args) {
      // Open input
      Scanner scan = new Scanner(System.in);

      // Read in number of test cases
      int T = scan.nextInt();
      for (int t = 1; t <= T; t++) {
         // Read in number of rows and columns
         R = scan.nextInt();
         C = scan.nextInt();

         // Read in the map
         map = new char[R][C];
         good = new boolean[R][C];
         for (int i = 0; i < R; i++)
            map[i] = scan.next().toCharArray();

         // Loop over the map and try all starting points of diamonds
         for (int i = 0; i < R; i++) {
            for (int j = 0; j < C; j++) {
               if (map[i][j] == '/')
                  check(i, j, 1, true);
            }
         }

         // Output the new slab
         System.out.println("Slab #" + t + ":");
         for (int i = 0; i < R; i++) {
            for (int j = 0; j < C; j++) {
               // If good, print the character, otherwise print a '.'
               System.out.print(good[i][j] ? map[i][j] : ".");
            }
            System.out.println();
         }
      }

      // Close input
      scan.close();
   }



   /*
       (a,b) is coordinate within the map matrix we are checking
       dist is the distance between the two characters we are checking
       top is a boolean representing whether we are checking for the top
           of a diamond
   */
   private static boolean check(int a, int b, int dist, boolean top) {
      // The only time dist < 0 will be when we've finished checking
      // a valid diamond
      if (dist < 0)
         return true;

      // This checks to see if the character we are checking is out of bounds
      if (a < 0 || a >= R || b < 0 || b >= C || b + dist >= C)
         return false;
      
      boolean valid = false;
      if (top && map[a][b] == '/' && map[a][b + dist] == '\\') {
         // Check if we can make a bigger diamond top of a diamond
         valid = check(a + 1, b - 1, dist + 2, true);

         // Check if we can form the bottom half of a diamond from here
         valid |= check(a + 1, b, dist, false);

         // If this is a good diamond, mark it
         if (valid)
            good[a][b] = good[a][b + dist] = true;
      }
      if (!top && map[a][b] == '\\' && map[a][b + dist] == '/') {
         if (valid = check(a + 1, b + 1, dist - 2, false))
            good[a][b] = good[a][b + dist] = true;
      }
      return valid;
   }
}
